client/lib/notification-settings-store/test/toggle-state.js

```diff
@@ -0,0 +1,364 @@
+/** @format */
+/**
+ * Internal dependencies
+ */
+import toggleState from '../toggle-state';
+
+describe( 'toggleState.wpcom', () => {
+	describe( 'when setting does not yet have a value', () => {
+		test( 'should assign the value of `true` to that setting', () => {
+			const settingToToggle = 'exampleSetting';
+			const startingState = {
+				settings: {},
+			};
+
+			const actual = toggleState.wpcom( startingState, null, null, settingToToggle );
+			const expected = {
+				settings: {
+					dirty: {
+						wpcom: {
+							[ settingToToggle ]: true,
+						},
+					},
+				},
+			};
+
+			expect( actual ).toEqual( expected );
+		} );
+	} );
+
+	describe( 'when setting already has a value', () => {
+		test( 'should toggle the value of that setting', () => {
+			const settingToToggle = 'exampleSetting';
+			const startingState = {
+				settings: {
+					dirty: {
+						wpcom: {
+							[ settingToToggle ]: true,
+						},
+					},
+				},
+			};
+
+			const actual = toggleState.wpcom( startingState, null, null, settingToToggle );
+			const expected = {
+				settings: {
+					dirty: {
+						wpcom: {
+							[ settingToToggle ]: false,
+						},
+					},
+				},
+			};
```

Comment:
An idea to make the tests less verbose and repetitive: create helper for creating a mock state:
```js
const startingState = createState( [ 'settings', 'dirty', 'wpcom', settingToToggle ], true );
```
and check the expected value by:
```js
expect( get( state, [ 'settings', 'dirty', 'wpcom', settingToToggle ] ) ).toBe( false );
```
extracting a `path` variable makes this particular test even shorter:
```js
const path = [ 'settings', 'dirty', 'wpcom', settingToToggle ];
```

Reply:
I tend to prefer avoiding doing any abstractions in the tests if it can be helped, instead opting for each test to show as clearly as possible and with as little indirection as possible what the input and output should be. I like the idea of `createState` somewhat but I'd still prefer to not have the future reader have to look at what `createState` is doing and how it relates to the test.

In terms of check with `expect( get( state, ... ) ).toBe( false );` I'd also err away from such a direct check. Reason being that our unit logic might change by mistake to add other things to the object. The `settingToToggle` might be successfully toggled but other side effects could sneak in unchecked.

That said, I've changed the responsibility of `toggle-state.js` a little and so now it returns the minimum object shape needed (i.e. no extra `settings.dirty.x`), so that's an improvement at least :)

Hope that makes sense and is reasonable?

Reply:
I think that tests should be written the same way as any other program. Avoid repetitive code, extract it to functions. Test the program behavior (did it toggle the setting?) rather than internal implementation (exact shape of the state object).

A good criterion for a well written test would be: if I make one-line change in the implementation, is the test update also an one-line change? Or do I need to update 50 repetitive tests?

I'm just offering a personal opinion here, no need to update the PR if you don't agree enthusiastically 🙂